### DISCUSSION TITLE
Ability to remove selected lists from generation

### DISCUSSION THREAD

**Discussion by nseq** [AIrjen/OneButtonPrompt#33] [Ideas]
Let's say I don't want anything from "cultures.csv" to appear in my prompt, would love to have an ability to do so. Right now I guess I can add something from that list so that other things wont appear from it, but sometimes I don't want anything from that list or any other to appear at all.

**Comment by AIrjen** [comment]
Thanks! In the background, in the latest update, I am already laying the groundwork in the code for this. ( as part of "smart subject" ).

I want to be able, to give more "control". This is where custom subject and now the list replacements and add-ons have come from.
I have already been thinking around the idea of being able to turn off/on specifics and able to set the random occurance. However, I also don't want to do too much, else other extensions are better suited, such as Dynamic Prompts.

In the latest update, you can already supply an replacement list, but that breaks if it is empty.

So short term:
- I can make it so list generation turns off, if the replacement list is empty (quick workaround)

Long term:
I want to have an interface, where you can turn things on and off and set the randomization levels.

**Reply by tusharbhutt** [reply]
Is it possible to add a custom file or two as well, one that won't be wiped out by an upgrade?  I have a 12K line prompt file from ChatGPT as well as many spreadsheets I made a few months ago that would work well with this.  I know I could add to the CSV files, but I'd like to keep it clean as possible

**Reply by AIrjen** [reply]
Hi, see the documentation surrounding custom files here:
https://github.com/AIrjen/OneButtonPrompt/blob/main/user_guides/custom_files.md

This has been built so you can maintain your own files and additions, without fear of them being replaced on an upgrade.

Also, if you have anything cool that you would suggest to be added, let me know!

**Reply by tusharbhutt** [reply]
That is awesome, I'll have a deeper look but it seems one would need to categorize keywords so I have my work cut out for me :)

**Reply by tusharbhutt** [reply]
> Hi, see the documentation surrounding custom files here: https://github.com/AIrjen/OneButtonPrompt/blob/main/user_guides/custom_files.md
> 
> This has been built so you can maintain your own files and additions, without fear of them being replaced on an upgrade.
> 
> Also, if you have anything cool that you would suggest to be added, let me know!

Hi, I do have a suggestion... when using the bath function, dynamic prompts can spit out 4 different random wildcard prompts if I choose a batch size of four.  With OBP, you also get four random outputs if you use a wildcard, but only for the wildcard text and not three rest of the OBP generated prompt

The same happens if you just run four in sequence without a wildcard.  You get four of the same prompt with different seeds.  It would be cool if there was an option to randomize on batch sizes greater than one as those runs save time

**Reply by AIrjen** [reply]
Hi,

people have been requesting this, but I can't seem to make this work. I have tried a couple of times now.

I am also trying to get this to work with Dynamic Prompts, but this ALSO gives me the same prompt for each Batch Size, and also only changes prompt/wildcards on Batch Count. So it seems Dynamic Prompts also doesn't support this.

Or am I missing a setting somewhere? I have tried with: 
{cat|dog}
or
__altar__

I tried with Combinatorial generation On and Off, and with Prompt Magic turned on and off.

So far each, extension I have used works the same. It uses the same prompt for each Batch Size, and a different one for Batch Count.
For X/Y/Z plot its the same.

**Reply by tusharbhutt** [reply]
> Hi,
> 
> people have been requesting this, but I can't seem to make this work. I have tried a couple of times now.

Thank you for trying, much appreciated :)

>  Or am I missing a setting somewhere? I have tried with: {cat|dog} or **altar**
 
> So far each, extension I have used works the same. It uses the same prompt for each Batch Size, and a different one for Batch Count. For X/Y/Z plot its the same.

I **think** SD loads each time it's run so if you used \_\_mylist\_\_ with batch size of four, you will get four different random choices from that list. 

For example, the prompt here was "professional digital airbrush art of \_\_animals\_\_" and run with batch count of two and batch size of two. Instead o getting two sets of two animals, there are four random ones:

![image](https://github.com/AIrjen/OneButtonPrompt/assets/26264462/e5d9d129-c60e-4f1e-96b6-af8689cdb9d3)

**Reply by AIrjen** [reply]
Ohhh noooo:

![grid-0001](https://github.com/AIrjen/OneButtonPrompt/assets/130234949/0dff98e0-d77d-4448-bd76-77ad7051a8fe)

It's a reverse, it works on my machine!

But...! I thank you for this, because I figured out what the problem was. It is the "Unprompted" extension which is interfering.
This is also what caused me to be never to solve this problem apparently.

So, I have two things to tell:
1. Did you know you can use the wildcards in the prefix/suffix and subject override fields in One Button Prompt?
2. I will start this research up again, but I'm in the middle of a huge update, which will take at least a week more to finish, maybe even two.

This will be added, but I can't tell you when.

**Reply by tusharbhutt** [reply]
> But...! I thank you for this, because I figured out what the problem was. It is the "Unprompted" extension which is interfering. This is also what caused me to be never to solve this problem apparently.
> 
Glad to help in some small way
> So, I have two things to tell:
> 
>     1. Did you know you can use the wildcards in the prefix/suffix and subject override fields in One Button Prompt?
> 
Yup, and it's really cool for what it does :)
>     2. I will start this research up again, but I'm in the middle of a huge update, which will take at least a week more to finish, maybe even two.

Awesome, looking forward to it!
> 
> This will be added, but I can't tell you when.

**Comment by AIrjen** [comment]
So I made a update yesterday with some internals reworked.

Now, any empty list provided as a "replace" will automatically not be used anymore.
Same goes for referencing empty lists via wildcards, these will be emptied.

During the rework, I removed a lot of hardcoded options, and put them into lists as well. They can be found under \csvfiles\special_lists\
These can also be added/replaced just as the other files.

Enjoy!

**Comment by AIrjen** [comment]
An update for someone who visits here, there is now a config file that can control (most)  these things as well:
https://github.com/AIrjen/OneButtonPrompt/blob/main/user_guides/config_file.md
